R/calcHDDCDD.R

```diff
@@ -1,101 +1,714 @@
+      }
       return(tmp)
-    }))
-  }))
-
-
-  # Smooth the trajectories
-  dataSmoothed <- lapply(ssps, function(scenario) {
-    lapply(rcps, function(rcp) {
-      thirdDim <- paste(scenario, variable, rcp, tlimit, sep = ".")
-      out <- lowpass(data[, , thirdDim], i = 100)
-      return(out)
-    })
-  })
-  dataSmoothed <- do.call("mbind", unlist(dataSmoothed, recursive = FALSE))
-
-
-  # Smooth of the transition between historical and scenario values
-  data <- data * lambda + dataSmoothed * (1 - lambda)
-
-  # replace negative values with 0 (large neg values only  for  small countries)
-  data[which(data < 0)] <- 0
-
-  # Set regions where there is either no value for history or for scenarios to 0
-  data["GRL", , ] <- 0
-  data[c("ASM", "FSM", "GUM", "MDV", "MHL", "TUV", "TKL", "SJM"), , ] <- 0
-
-  # Add the no Climate Change scenario
-  noC <- data[, , "rcp2p6"]
-
-  # Fill with the mean of the five last periods
-  noC[, yearsScenario, ] <-
-    dimSums(noC[, c((max(yearsHistory) - 4):max(yearsHistory)), ], dim = 2) / 5
-  getNames(noC) <- gsub("rcp2p6", "rcpNoC", getNames(noC))
-  data <- mbind(data, noC)
-
-  pop <- calcOutput("Population", aggregate = FALSE, FiveYearSteps = FALSE)
-  getNames(pop) <- tolower(sub("pop_(....).?$", "\\1", getNames(pop)))
-  pop <- pop[, , ssps]
-  y <- intersect(getYears(data), getYears(pop))
-  data <- data[, y, ]
-  pop <- pop[, y, ]
-
-  return(list(x = data,
-              weight = pop,
-              unit = "dK/yr",
-              description = "Heating degree days or cooling degree days"))
+      },
+      USE.NAMES = TRUE
+    )
+    return(baitInput)
+  }
+
+
+  #--- CALCULATE BAIT (building-adjusted internal temperature)
+
+  # NOTE: The parameters are taken from Staffell et al. 2023
+
+  # counterfactuals for solar, wind, humidity, temperature
+  cfac <- function(t, type, params=NULL) {
+    if (is.null(params)) {
+      params <- switch(type,
+                       s = c(100, 7),
+                       w = c(4.5, -0.025),
+                       h = c(1.1, 0.06),
+                       t = c(16))}
+
+    print(names(params))
+    if      (type == "s") {return(params[[1]] + params[[2]]*t)}
+    else if (type == "w") {return(params[[1]] + params[[2]]*t)}
+    else if (type == "h") {return(exp(params[[1]] + params[[2]]*t))}
+    else if (type == "t") {return(params[[1]])}
+
+    else {print("No valid parameter type specified.")}
+  }
+
+
+  # smooth data
+  smooth <- function(r, weight) {
+    # smooth bait over preceding two days with smoothing parameter sigma
+    print("smooth")
+
+    # one day indented
+    r1D <- r[[c(nlyr(r), 1:(nlyr(r) - 1))]]
+    r1D[[1]] <- 0
+
+    # two days indented
+    r2D <-  r[[c(nlyr(r)-1, nlyr(r), 1:(nlyr(r) - 2))]]
+    r2D[[1:2]] <- 0
+
+    # smooth
+    rSmooth <- (r + weight[[4]]*r1D + weight[[4]]**2 * r2D) / (1 + weight[[4]] + weight[[4]]**2)
+
+    return(rSmooth)
+  }
+
+  blend <- function(baitDF, temp, weight) {
```

Comment:
Please explain the sigmoid function and the reasoning for blending outdoor temperature and BAIT.